doc/sigsh.1
.\"	Copyright (c) 2010-2016 Yahoo! Inc.
.\"
.\"	This manual page was originally written by Jan Schaumann
.\"	<jschauma@yahoo-inc.com> in September 2010.
.Dd February 08, 2016
.Dt SIGSH 1
.Os
.Sh NAME
.Nm sigsh
.Nd a signature verifying shell
.Sh SYNOPSIS
.Nm
.Op Fl Vdx
.Op Fl f Ar certs
.Op Fl p Ar prog
.Sh DESCRIPTION
.Nm
is a non-interactive, signature requiring and verifying command
interpreter.
More accurately, it is a signature verification wrapper around a given
shell.
It reads input in PKCS#7 format from standard in, verifies the signature
and, if the signature matches, pipes the decoded input into the command
interpreter.
.Sh OPTIONS
.Nm
supports the following flags:
.Bl -tag -width s_shell_
.It Fl V
Report version information and exit.
.It Fl d
Don't execute the commands, just show them.
.It Fl f Ar certs
Read ceritificates to trust from this file.
.It Fl p Ar prog
Pipe commands into this interpreter instead of the default
.Xr bash 1 .
.It Fl x
Enable debugging (mnemomic 'xtrace', as
.Xr sh 1 Ns ).
.El
.Sh DETAILS
Conceptually similar to Microsoft Windows' Powershell ExecutionPolicy (as
set to 'allSigned'),
.Nm
will only execute any commands from the input if a valid signature is
found.
This allows, for example, a headless user to be able to run any arbitrary
set of commands (if provided by trusted entities) without having to give
it a fully interactive login shell.
By specifying a different interpreter to which to pass the verified input,
.Nm
can be used for almost anything requiring input verification so long as
the tool invoked accepts input from standard in.
.Pp
.Nm
is intentionally kept as simple as possible and does not provide for a
whole lot of customization via either a startup file or any command-line
options.
.Sh INPUT
.Nm
reads input from standard in.
That is, unlike other interactive command interpreters, it cannot be
invoked from the terminal to read commands one at a time.
.Nm
relies on (and shells out to)
.Xr openssl 1
for signature verification.
In particular, it expects input to be in PKCS#7 format, containing signed
data to be passed to the command interpreter.
In order to verify the signature,
.Nm
needs to have available a matching certificate (see section FILES).
.Sh OUTPUT
By default,
.Nm
does not generate any output itself.
If input verification fails, it will return an error code (see section
EXIT STATUS) and print a brief message to STDERR; otherwise, it will pipe
the validated input to the given command interpreter, letting it generate
any and all output (both to standard out and standard error).
.Sh EXAMPLES
The following examples illustrate possible usage of this tool.
.Pp
To execute the commands in the file 'script.bash':
.Bd -literal -offset indent
openssl smime -sign -nodetach -signer mycert.pem -inkey mykey.pem \\
    -in script.bash -outform pem | sigsh
.Ed
.Pp
To execute the perl code contained in the signed PKCS#7 file 'code.pem':
.Bd -literal -offset indent
sigsh -p /usr/bin/perl <code.pem
.Ed
.Sh EXIT STATUS
.Nm
will exit with the rather unusual return code of 127 if verification of
the input fails (for whatever reason).
Otherwise, it will return the exit code of the interpreter invoked.
.Sh ENVIRONMENT
.Nm
clears the environment before passing the verified input on to the
interpreter.
Therefor, the input must make sure to explicitly set any variables it may
rely on.
.Sh FILES
.Nm
uses the following files:
.Bl -tag -width _etc_sigsh_pem_
.It /etc/sigsh.pem
The file containing all certificates that
.Nm
will verify the input against.
.El
.Sh SEE ALSO
.Xr openssl 1 ,
.Xr smime 1
.Sh HISTORY
.Nm
was originally written by
.An Jan Schaumann
.Aq jschauma@yahoo-inc.com
in September 2010.
.Sh BUGS
Please report bugs and feature requests to the author.


CHANGES
2015-02-10:
 * merge changes from Y!:
   * change '-c' to '-f', since login shells might get invoked with '-c',
     which opens up the possibility of allowing a user to provide their
     own certificate via 'ssh host -- /dev/tty'
   * correct error reporting, since verifyArgs() is executed in a subshell
   * quote args to verifyArgs

2011-08-07:
 * significantly increase performance by only handling input line-by-line
   if we are in trace mode


src/sigsh.sh
#! /bin/sh
#
# Copyright (c) 2010,2011,2016 Yahoo! Inc.
# All rights reserved.
#
# Originally written by Jan Schaumann <jschauma@yahoo-inc.com> in October
# 2010.
#
# sigsh is a non-interactive, signature requiring and verifying command
# interpreter. More accurately, it is a signature verification wrapper
# around a given shell. It reads input in PKCS#7 format from standard in,
# verifies the signature and, if the signature matches, pipes the decoded
# input into the command interpreter.
#
# Redistribution and use of this software in source and binary forms,
# with or without modification, are permitted provided that the following
# conditions are met:
#
# * Redistributions of source code must retain the above
#   copyright notice, this list of conditions and the
#   following disclaimer.
#
# * Redistributions in binary form must reproduce the above
#   copyright notice, this list of conditions and the
#   following disclaimer in the documentation and/or other
#   materials provided with the distribution.
#
# * Neither the name of Yahoo! Inc. nor the names of its
#   contributors may be used to endorse or promote products
#   derived from this software without specific prior
#   written permission of Yahoo! Inc.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
# IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
# TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
# PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
# Yes, the author is aware that we have more lines of comments than actual
# code, since the latter can be condensed into a single command like
#
# openssl smime -verify -inform pem -CAfile mycert.pem

PATH="/bin:/usr/bin:/sbin/:/usr/sbin"

###
### Globals
###
CERTS="/etc/sigsh.pem"
PROGRAM="bash"
SSL_VERIFY="openssl smime -verify -inform pem -CAfile"
PROGNAME="${0##*/}"
VERSION="0.7.1"
XTRACE=0

###
### Functions
###

# function : error
# purpose  : print given message to STDERR
# inputs   : msg

error() {
	local msg="$@"

	echo "${PROGNAME}: $msg" >&2
}

# function : usage
# purpose  : print usage statement

usage() {
	cat <<EOH
Usage: ${PROGNAME} [-Vdx] [-f certs] [-p program]
         -V          Print version info and exit.
         -d          Don't execute commands.
         -f certs    Read certs to trust from this file.
         -p program  Pipe commands into 'program'.
         -x          Enabled debugging.
EOH
}

# function : verifyArg
# purpose  : ensure given arg is sane for shell evaluation by matching it
#            against a simple restrictive RE
# inputs   : a string
# prints   : the given string if it matches
# returns  : 0 on success, 1 on invalid input

verifyArg() {
	local arg="${1}"

	if expr "${arg}" : "[a-zA-Z0-9/_.-]*$" >/dev/null 2>&1 ; then
		echo "${arg}"
		return 0
	fi

	error "Argument must match ^[a-zA-Z0-9/_.-]*$."
	return 1
}

# function : xtrace
# purpose  : print given message prepended with "+ " if xtrace is set
# inputs   : a string

xtrace() {
	local msg="$1"

	if [ ${XTRACE} -gt 0 ]; then
		echo "+ ${msg}" >&2
	fi
}

###
### Main
###

while getopts 'Vdf:p:x' opt; do
	case ${opt} in
		V)
			echo "${VERSION}"
			exit 0
			# NOTREACHED
		;;
		d)
			PROGRAM="cat"
		;;
		f)
			CERTS=$(verifyArg "${OPTARG}")
			[ $? -gt 0 ] && exit 1
		;;
		p)
			PROGRAM=$(verifyArg "${OPTARG}")
			[ $? -ne 0 ] && exit 1
		;;
		x)
			XTRACE=1
		;;
		*)
			usage
			exit 1
			# NOTREACHED
		;;
	esac
done
shift $(($OPTIND - 1))

if [ $# -gt 0 ]; then
	usage
	exit 1
	# NOTREACHED
fi

verify="${SSL_VERIFY} ${CERTS}"
xtrace "${verify}"

if [ ${XTRACE} -eq 0 ]; then
	verify="${verify} 2>/dev/null"
fi

# We could just do something like
#
# eval ${verify} | while read -r line; do
#
# but that would return 0 even if the verification failed.  If we want to
# return the special exit code on verification failure, we need to keep
# the whole output in memory, but that seems reasonable compared to the
# unreadable shenanigans we'd otherwise have to play with shell redirection
# etc.  Ditching the special exit code is the other option, but I think
# it'd be useful to be able to discern verification failure based on $?.

output=$(eval ${verify})
if [ $? -gt 0 ]; then
	echo "Unable to verify given input." >&2
	exit 127
fi

IFS='
'
# We only loop through this line by line if we're tracing.  Otherwise,
# this would be a significant performance penalty for large input scripts.
if [ ${XTRACE} -gt 0 ]; then
	for line in ${output}; do
		xtrace "${line}"
		echo "${line}" | tr -d '\r'
	done | ${PROGRAM}
else
	echo "${output}" | tr -d '\r' | ${PROGRAM}
fi
